File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFontLoaderModule.java
package versioned.host.exp.exponent.modules.universal;

import android.content.Context;

import org.unimodules.core.InvalidArgumentException;
import org.unimodules.core.Promise;
import org.unimodules.core.errors.CodedException;
import org.unimodules.interfaces.constants.ConstantsInterface;
import org.unimodules.interfaces.filesystem.FilePermissionModuleInterface;
import org.unimodules.interfaces.filesystem.Permission;

import expo.modules.font.FontLoaderModule;

public class ScopedFontLoaderModule extends FontLoaderModule {
  private static final String SCOPED_FONT_PREFIX = "Expo-";

  public ScopedFontLoaderModule(Context context) {
    super(context);
  }

  @Override
  public void loadAsync(String providedFontFamilyName, String localUri, Promise promise) {
    String fontFamilyName = providedFontFamilyName;

    if (isScoped()) {
      // Validate font family name before we prefix it
      if (providedFontFamilyName == null) {
        promise.reject(new InvalidArgumentException("Font family name cannot be empty (null received)"));
        return;
      }

      // Scope font family name
      fontFamilyName = SCOPED_FONT_PREFIX + providedFontFamilyName;

      // Ensure filesystem access permissions
      FilePermissionModuleInterface filePermissionModule = mModuleRegistry.getModule(FilePermissionModuleInterface.class);
      if (filePermissionModule != null) {
        if (!filePermissionModule.getPathPermissions(getContext(), localUri).contains(Permission.READ)) {
          promise.reject(new LocationAccessUnauthorizedError(localUri));
          return;
        }
      }
    }

    super.loadAsync(fontFamilyName, localUri, promise);
  }

  private boolean isScoped() {
    ConstantsInterface constantsModule = mModuleRegistry.getModule(ConstantsInterface.class);
    // If there's no constants module, or app ownership isn't "expo", we're not in Expo Client.
    return constantsModule != null && "expo".equals(constantsModule.getAppOwnership());
  }

  public static class LocationAccessUnauthorizedError extends CodedException {
    public LocationAccessUnauthorizedError(String uri) {
      super("You aren't authorized to load font file from: " + uri);
    }

    @Override
    public String getCode() {
      return "ERR_LOCATION_ACCESS_UNAUTHORIZED";
    }
  }
}

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFontLoaderModule.kt
package versioned.host.exp.exponent.modules.universal

import android.content.Context
import android.net.Uri
import expo.modules.font.FontLoaderModule
import org.unimodules.core.Promise
import org.unimodules.core.errors.CodedRuntimeException
import org.unimodules.core.errors.InvalidArgumentException
import org.unimodules.interfaces.constants.ConstantsInterface
import org.unimodules.interfaces.filesystem.FilePermissionModuleInterface
import org.unimodules.interfaces.filesystem.Permission

class ScopedFontLoaderModule(context: Context?) : FontLoaderModule(context) {
  override fun loadAsync(fontFamilyName: String?, localUri: String?, promise: Promise) {
    var resolvedFontFamilyName = fontFamilyName
    if (isScoped) {
      // Validate font family name before we prefix it
      if (resolvedFontFamilyName == null) {
        promise.reject(InvalidArgumentException("Font family name cannot be empty (null received)"))
        return
      }

      // Scope font family name
      resolvedFontFamilyName = "$SCOPED_FONT_PREFIX$resolvedFontFamilyName"

      // Ensure filesystem access permissions
      val filePermissionModule = mModuleRegistry?.getModule(FilePermissionModuleInterface::class.java)
      if (filePermissionModule != null) {
        val localFontPath = Uri.parse(localUri).path
        if (localFontPath == null) {
          promise.reject(InvalidArgumentException("Could not parse provided local font URI as a URI with a path component."))
          return
        }
        if (!filePermissionModule.getPathPermissions(context, localFontPath).contains(Permission.READ)) {
          promise.reject(LocationAccessUnauthorizedError(localFontPath))
          return
        }
      }
    }
    super.loadAsync(resolvedFontFamilyName, localUri, promise)
  }

  private val isScoped: Boolean
    get() {
      val appOwnership = mModuleRegistry?.getModule(ConstantsInterface::class.java)?.appOwnership
      // If there's no constants module, or app ownership isn't "expo", we're not in Expo Client.
      return appOwnership != null && "expo" == appOwnership
    }

  class LocationAccessUnauthorizedError(uri: String) : CodedRuntimeException("You aren't authorized to load font file from: $uri") {
    override fun getCode(): String {
      return "ERR_LOCATION_ACCESS_UNAUTHORIZED"
    }
  }

  companion object {
    private const val SCOPED_FONT_PREFIX = "ExpoFont-"
  }
}
